name: next
on: push
jobs:

  python312:
    runs-on: ubuntu-22.04
    strategy:
      fail-fast: false
    env:
      PATH: /github/home/.local/bin:/github/home/.cargo/bin:/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
      CC: "gcc"
      CFLAGS: "-O2 -fno-plt"
      LDFLAGS: "-O2 -Wl,--as-needed"
      CARGO_UNSTABLE_SPARSE_REGISTRY: "true"
      CARGO_PROFILE_RELEASE_DEBUG: "2"
      CARGO_PROFILE_RELEASE_DEBUG_ASSERTIONS: "true"
      CARGO_PROFILE_RELEASE_LTO: "false"
      CARGO_PROFILE_RELEASE_OVERFLOW_CHECKS: "true"
      CARGO_PROFILE_RELEASE_PANIC: "unwind"
    container:
      image: registry.hub.docker.com/library/python:3.12-rc-bullseye
      options: --user 0
    steps:
    - run: curl https://sh.rustup.rs -sSf | sh -s -- --default-toolchain nightly-2023-05-07 --profile minimal -y
    - run: python3 -m pip install --user --upgrade pip "maturin>=1,<2" wheel
    - uses: actions/checkout@v3

    - name: build
      run: |
        maturin build --release \
          --out=dist \
          --features=unstable-simd,yyjson \
          --compatibility off \
          --interpreter python${{ matrix.python.version }} \
          --target=x86_64-unknown-linux-gnu

    - run: python3 -m pip install --user dist/orjson*.whl
    - run: python3 -m pip install --user -r test/requirements.txt -r integration/requirements.txt

    - name: pytest
      continue-on-error: true
      timeout-minutes: 5
      run: pytest -s -rxX -v test

    - run: ./integration/run thread
      continue-on-error: true
      timeout-minutes: 2

    - run: ./integration/run http
      continue-on-error: true
      timeout-minutes: 2

    - run: ./integration/run init
      continue-on-error: true
      timeout-minutes: 2
